Document compiler and JVM args within Advanced GradleRIO (#2393)
wpilibsuite/frc-docs: source/docs/software/advanced-gradlerio/compiler-args.rst

Using Compiler Arguments
============================

> Compiler arguments allow us to change the behavior of our compiler. This includes making warnings into errors, ignoring certain warnings and choosing optimization level. Normally we would pass them in as flags when compiling our code but this doesn't work in GradleRIO. Instead we have to modify our build.gradle.

Reviewer: I would add a link to where the default flags are located

Author: where would one find the default flags?

Reviewer: https://github.com/wpilibsuite/native-utils/blob/0055bc77da9d9d37e0876d04af53b7317b78174f/src/main/java/edu/wpi/first/nativeutils/WPINativeUtilsExtension.java#L37

Author: There are a fair few there so I am just adding a link to them for readibility

C++
---

Platforms
^^^^^^^^^

> In order to avoid breaking different platforms with compiler flags we configure everything per platform. The platforms that are supported are :

Reviewer: I would lead in by saying different compilers on different platforms use different flags, and then say to avoid breaking

- windowsx86-64
- linuxathena

Configuring for a Platform
^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: groovy

  nativeUtils.platformConfigs.named('windowsx86-64').configure {
    it.cppCompiler.args.add("/utf-8")
  }

We can break this code up step by step. First we use nativeUtils to configure the platform which here is windowsx86-64, but you can replace for whichever platform you choose to configure. Then we take the C++ compiler and add an argument "/utf-8".

Java
----

> We can configure arguments to the JVM when compiling projects with GradleRIO. We can go into our build.gradle and do something similar to this :

Reviewer: clarify that the FRCJavaArtifact block already exists in the project, so you just have to add the jvmArgs line

.. code-block:: groovy

  frcJava(getArtifactClass('FRCJavaArtifact')) {

>     jvmArgs.add("DisableExplicitGC")

Reviewer: Does this actually work? I expected it to need `-XX:+DisableExplicitGC`

Author: tested that [here](https://github.com/spacey-sooty/JVM-Args-Testing/tree/master)

Reviewer: Did you deploy to a roboRIO?

Author: No just built locally

Reviewer: Those are only used when deploying to the roboRIO

Author: Ahh ok I will change it to your one

Using Compiler Arguments

> ============================

Reviewer 2: Title underlines should match the title

Compiler arguments allow us to change the behavior of our compiler. This includes making warnings into errors, ignoring certain warnings and choosing optimization level. Normally we would pass them in as flags when compiling our code but this doesn't work in GradleRIO. Instead we have to modify our build.gradle.

C++
---

Platforms
^^^^^^^^^

Different compilers on different platforms use a variety of different flags so in order to avoid breaking different platforms with compiler flags we configure everything per platform. The platforms that are supported are :

- windowsx86-64
- windowsx86
- windowsarm64

> - linuxathena

Reviewer: I think most are self explanatory, but this should be stated to be roborio

> Compiler arguments allow us to change the behavior of our compiler. This includes making warnings into errors, ignoring certain warnings and choosing optimization level. When compiling code a variety of flags are already included by default which can be found `here <https://github.com/wpilibsuite/native-utils/blob/0055bc77da9d9d37e0876d04af53b7317b78174f/src/main/java/edu/wpi/first/nativeutils/WPINativeUtilsExtension.java#L37>`. Normally it could be proposed that the solution is to pass them in as flags when compiling our code but this doesn't work in GradleRIO. Instead modify the build.gradle. **Warning Modifying arguments is dangerous and can cause unexpected behavior.**

Reviewer: use a rst `.. warning::`

> Different compilers on different platforms use a variety of different flags so in order to avoid breaking different platforms with compiler flags we configure everything per platform. The platforms that are supported are :

Reviewer 2: Run-on sentence

> - linuxarm32

Reviewer 2: I think there is a gradle file we can link instead of duplicating here. @ThadHouse

Author: I got that from looking through native utils. I can try and find the file where they were listed

Author: found [here](https://github.com/wpilibsuite/native-utils/blob/5b6aca5efed38188561ec26b58bebbe159e5f340/src/main/java/edu/wpi/first/nativeutils/WPINativeUtilsExtension.java#L97)

Author: Should I link this instead @Daltz333

- linuxarm64
- linuxx86-64
- osxuniversal

Configuring for a Platform
^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: groovy

  nativeUtils.platformConfigs.named('windowsx86-64').configure {
    it.cppCompiler.args.add("/utf-8")
  }

We can break this code up step by step. First we use nativeUtils to configure the platform which here is windowsx86-64, but you can replace for whichever platform you choose to configure. Then we take the C++ compiler and add an argument "/utf-8".

Java
----

> We can configure arguments to the JVM when compiling projects with GradleRIO. We can go into our build.gradle and do something similar to this(note the FRCJavaArtifact already exists you just need to add into it) :

Reviewer 2: I honestly don't know what this paragraph means. Why should I add Java compiler args (remote debugging/memory analysis)?

Additionally, we should add a warning to the top of this page that modifying arguments is dangerous and can lead to unexpected behavior.

Using Compiler Arguments
========================

> Compiler arguments allow us to change the behavior of our compiler. This includes making warnings into errors, ignoring certain warnings and choosing optimization level. When compiling code a variety of flags are already included by default which can be found `here <https://github.com/wpilibsuite/native-utils/blob/0055bc77da9d9d37e0876d04af53b7317b78174f/src/main/java/edu/wpi/first/nativeutils/WPINativeUtilsExtension.java#L37>`. Normally it could be proposed that the solution is to pass them in as flags when compiling our code but this doesn't work in GradleRIO. Instead modify the build.gradle.

Reviewer: ```suggestion
Compiler arguments allow us to change the behavior of our compiler. This includes making warnings into errors, ignoring certain warnings and choosing optimization level. When compiling code a variety of flags are already included by default which can be found `here <https://github.com/wpilibsuite/native-utils/blob/0055bc77da9d9d37e0876d04af53b7317b78174f/src/main/java/edu/wpi/first/nativeutils/WPINativeUtilsExtension.java#L37>__`. Normally it could be proposed that the solution is to pass them in as flags when compiling our code but this doesn't work in GradleRIO. Instead modify the build.gradle.
```

Reviewer: There's still a problem with the link

> .. warning:: Warning Modifying arguments is dangerous and can cause unexpected behavior.

Reviewer: Needs a new line before
```suggestion

.. warning:: Warning Modifying arguments is dangerous and can cause unexpected behavior.
```

Using Compiler Arguments
========================

Compiler arguments allow us to change the behavior of our compiler. This includes making warnings into errors, ignoring certain warnings and choosing optimization level. When compiling code a variety of flags are already included by default which can be found `here <https://github.com/wpilibsuite/native-utils/blob/0055bc77da9d9d37e0876d04af53b7317b78174f/src/main/java/edu/wpi/first/nativeutils/WPINativeUtilsExtension.java#L37>__`. Normally it could be proposed that the solution is to pass them in as flags when compiling our code but this doesn't work in GradleRIO. Instead modify the build.gradle.

> .. warning:: Warning Modifying arguments is dangerous and can cause unexpected behavior.

Reviewer 2: ```suggestion
.. warning:: Modifying arguments is dangerous and can cause unexpected behavior.
```

C++
---

Platforms
^^^^^^^^^

Different compilers and different platforms use a variety of different flags. Therefore to avoid breaking different platforms with compiler flags configure all flags per platform. The platforms that are supported are :

- windowsx86-64
- windowsx86
- windowsarm64
- linuxathena (roboRIO)
- linuxarm32
- linuxarm64
- linuxx86-64
- osxuniversal

Configuring for a Platform
^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: groovy

  nativeUtils.platformConfigs.named('windowsx86-64').configure {
    it.cppCompiler.args.add("/utf-8")
  }

> Lets break this code up step by step. First use nativeUtils to configure the platform which here is windowsx86-64, but can be replaced for whichever platform chosen to configure. Then take the C++ compiler and add an argument eg. "/utf-8".

Reviewer 2: ```suggestion
nativeUtils is used to configure the platform, in this case, `windowsx86-64`. This can be replaced for any platform listed in the previous section. Then arguments, such as `/utf-8` is appended to the C++ compiler.
```

Java
----

> We can configure arguments to the JVM when compiling projects with GradleRIO. These arguments can serve a variety of purposes for example remote debugging or memory analysis. To do this we can go into our build.gradle and do something similar to this(note the FRCJavaArtifact already exists just add to it) :

Reviewer 2: ```suggestion
Arguments can also be configured for Java. This can be accomplished by editing `build.gradle` and appending arguments to the `FRCJavaArtifact`. An example of this is shown below.
```

.. code-block:: groovy

  frcJava(getArtifactClass('FRCJavaArtifact')) {

>     jvmArgs.add("-XX:+DisableExplicitGC")

Reviewer 2: I'd rather not use disabling the GC as the argument here. Maybe instead raise the maximum ram limits?

Author: How would I do this? I am not super familiar with the Java side of things

Reviewer: This doesn't disable the GC. It instead disables users calls to the gc (e.g. `System.gc()` which isn't really recommended to do any way, so I'm not sure it's the worst thing to use as an example.

Other options would be `-XX:+PrintGCDetails` but you'd have to make sure it's printed somewhere useful

Author: In that case this seems like a decent option for an example I'll leave it

  }

> Lets go through this step by step. First enter the FRCJavaArtifact. This is the Java artifact deployed to the roboRIO. Here add an argument to our JVM using jvmArgs.add.

Reviewer 2: I think this is unnecessary.
